RHDEVDOCS-4601-how-to-resize-prometheus-storage-volumes (#52277)
openshift/openshift-docs: modules/monitoring-resizing-a-persistent-storage-volume.adoc

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: PROCEDURE
[id="resizing-a-persistent-storage-volume_{context}"]
= Resizing a persistent storage volume

> {product-title} does not support resizing an existing persistent storage volume using StatefulSets, even though the `StorageClass` setting for a custom resource (CR) supports persistent volume sizing.

Author: changed per your suggestion

> Therefore, even if you update the `storage` field for an existing persistent volume claim (PVC) with a larger size, these settings are not automatically propagated to the associated persistent volume (PV).

Author: changed per your suggestion

> However, resizing a PV is still possible by using a manual process. If you want to resize a PV for a monitoring component such as Prometheus, Thanos Ruler, or Alertmanager, you can update the appropriate config map for the Cluster Monitoring Operator (CMO), patch the PVC, and then delete and orphan the pods.

Reviewer: ```suggestion
However, resizing a PV is still possible by using a manual process. If you want to resize a PV for a monitoring component such as Prometheus, Thanos Ruler, or Alertmanager, you can update the appropriate config map where the component is configured, patch the PVC, and then delete and orphan the pods.
```

Author: changed per your suggestion

+
[source,terminal]
----
$ oc -n openshift-user-workload-monitoring edit configmap user-workload-monitoring-config
----

.. Update the PVC configuration for the monitoring component under `data/config.yaml`:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    <component>:
      volumeClaimTemplate:
        spec:
          storageClassName: <storage_class>
          resources:
            requests:
              storage: <amount_of_storage> <1>
----
<1> Specify the new size for the storage volume.
+
The following example changes the PVC size to 100 gigabytes the Prometheus instance that monitors user-defined projects:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    prometheus:
      volumeClaimTemplate:
        spec:
          storageClassName: local-storage
          resources:
            requests:
              storage: 100Gi
----
+

> The following example changes the PVC size to 20 gigabytes for Thanos Ruler:

Reviewer: Just would like to mention that users could also resize the size of Alermanager and Prometheus for UWM, not sure if we want to add examples  for that (personally I don't see the need) but just mentioning it

Author: I added a note about this at the beginning of this section of the procedure

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    thanosRuler:
      volumeClaimTemplate:
        spec:
          storageClassName: *local-storage*
          resources:
            requests:
              storage: 20Gi
----
+
[NOTE]
====
Storage requirements for the `thanosRuler` component depend on the number of rules that are evaluated and how many samples each rule generates.
====

> . Save the file to apply the changes. The pods affected by the new configuration restart automatically.

Reviewer: I don't think any pods will restart after saving the file, since in this procedure what we will see are new pods spawning once we patch the existing ones
```suggestion
. Save the file to apply the changes.
```

Reviewer: This is not correct, the pods will indeed re-start

Author: Ok. Not making any changes here

+

> [WARNING]

Reviewer: Not sure if this warning here makes sense

====
When you save changes to a monitoring config map, the pods and other resources in the related project might be redeployed. The running monitoring processes in that project might also be restarted.
====

> . Manually patch every PVC with the updated storage request. The following example resizes the storage size for the Prometheus component to 100Gi:

Reviewer: See suggestion bellow
```suggestion
. Manually patch every PVC with the updated storage request. The following example resizes the storage size for the `openshift-monitoring` Prometheus component to 100Gi:
```

Author: changed per your suggestion

+
[source,terminal]
----

> # for p in $(oc get pvc -l prometheusk8s/name=example -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \

Reviewer: I think it would be better to provide an example that would apply to OCP I have to doublecheck this on the cluster but I think like this it would actually patch the prometheus pvc
```suggestion
# for p in $(oc get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \
```

Reviewer: I've tested and it works with this label

Author: changed the label

  oc patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}'; \
  done
----

. Delete the underlying StatefulSet with the `--cascade=orphan` parameter:
+
[source,terminal]
----

> # oc delete statefulset -l prometheusk8s/name=example --cascade=orphan

Reviewer: ```suggestion
# oc delete statefulset -l app.kubernetes.io/name=prometheus --cascade=orphan
```

Author: changed

[id="resizing-a-persistent-storage-volume_{context}"]
= Resizing a persistent storage volume

{product-title} does not support resizing an existing persistent storage volume using StatefulSets, even if the underlying `StorageClass` used supports persistent volume sizing.
Therefore, even if you update the `storage` field for an existing persistent volume claim (PVC) with a larger size, this setting will not be propagated to the associated persistent volume (PV).

However, resizing a PV is still possible by using a manual process. If you want to resize a PV for a monitoring component such as Prometheus, Thanos Ruler, or Alertmanager, you can update the appropriate config map where the component is configured, patch the PVC, and then delete and orphan the pods.
Orphaning the pods recreates the StatefulSet immediately and automatically updates the size of the volumes mounted in the pods with the new PVC settings.
No service disruption occurs during this process. 

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* *If you are configuring core {product-title} monitoring components*:
** You have access to the cluster as a user with the `cluster-admin` role.

> ** You have created the `cluster-monitoring-config` `ConfigMap` object.

Reviewer 2: since we are talking about `Resizing a persistent storage volume` here, we'd better mention we've already attached PVs for components,and we want to resize the storage size, so we'd better add one pre-condition:
`You have configured PVCs for core OpenShift Container Platform monitoring components`

Author: Added `You have configured at least one PVC for core {product-title} monitoring components.`

* *If you are configuring components that monitor user-defined projects*:

> ** You have access to the cluster as a user with the `cluster-admin` role, or as a user with the `user-workload-monitoring-config-edit` role in the `openshift-user-workload-monitoring` project.

Reviewer 2: add
`A cluster administrator has enabled monitoring for user-defined projects.`

> ** You have created the `user-workload-monitoring-config` `ConfigMap` object.

Reviewer 2: since we are talking about Resizing a persistent storage volume here, so we'd better add one pre-condition:
`You have configured PVCs for components that monitor user-defined projects`

Author: Added `You have configured at least one PVC for components that monitor user-defined projects.`

.Procedure

. Edit the `ConfigMap` object:

> ** *To configure a PVC for a component that monitors core {product-title} projects*:

Reviewer 2: since we are talking about Resizing a persistent storage volume here
To configure a PVC
change to below is better
`To resize a PVC`

Author: changed

.. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

.. Add a new storage size for the PVC configuration for the component under `data/config.yaml`:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    <component>: <1>
      volumeClaimTemplate:
        spec:
          storageClassName: <storage_class> <2>
          resources:
            requests:
              storage: <amount_of_storage> <3>
----
<1> Specify the core monitoring component.
<2> Specify the storage class.

> <3> Specify the new size for the storage volume.

Reviewer 2: Specify the size for the storage volume.

+

> The following example configures a PVC that changes the local persistent storage to 100 gigabytes for the Prometheus instance that monitors core {product-title} components:

Reviewer 2: that changes the local persistent storage
->
that uses the local persistent storage

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      volumeClaimTemplate:
        spec:
          storageClassName: local-storage
          resources:
            requests:
              storage: 100Gi
----
+

> The following example changes a PVC that changes local persistent storage for Alertmanager to 40 gigabytes:

Reviewer 2: The following example **changes** a PVC that changes local persistent storage for Alertmanager to 40 gigabytes
->
The following example **configures** a PVC that changes local persistent storage for Alertmanager to 40 gigabytes

Author: changed

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    alertmanagerMain:
      volumeClaimTemplate:
        spec:

>           storageClassName: *local-storage*

Reviewer 2: storageClassName: local-storage

Author: fixed

          resources:
            requests:
              storage: 40Gi
----

> ** *To configure a PVC for a component that monitors user-defined projects*:

Reviewer 2: since we are talking about Resizing a persistent storage volume here
`To configure a PVC`
change to below is better
`To resize a PVC`

Author: changed

+
[NOTE]
====

> You can resize the volumes for the Alertmanager, Thanos Ruler and Prometheus instances that monitor user-defined projects.

Reviewer 2: dedicated Alertmanager for user workload monitoring is since 4.11, by default, the dedicated Alertmanager is not enabled , since this doc is for 4.9+, we don't need to mention Alertmanager here

====
+
.. Edit the `user-workload-monitoring-config` `ConfigMap` object in the `openshift-user-workload-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-user-workload-monitoring edit configmap user-workload-monitoring-config
----

.. Update the PVC configuration for the monitoring component under `data/config.yaml`:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    <component>: <1>
      volumeClaimTemplate:
        spec:
          storageClassName: <storage_class> <2>
          resources:
            requests:
              storage: <amount_of_storage> <3>
----
<1> Specify the core monitoring component.
<2> Specify the storage class.

> <3> Specify the new size for the storage volume.

Reviewer 2: Specify the size for the storage volume.

+

> The following example changes the PVC size to 100 gigabytes the Prometheus instance that monitors user-defined projects:

Reviewer 2: changes the PVC size to 100 gigabytes
->
sets the PVC size to 100 gigabytes

> The following example changes the PVC size to 100 gigabytes the Prometheus instance that monitors user-defined projects:

Reviewer 2: The following example changes the PVC size to 100 gigabytes for the Prometheus instance that monitors user-defined projects

Author: fixed

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    prometheus:
      volumeClaimTemplate:
        spec:
          storageClassName: local-storage
          resources:
            requests:
              storage: 100Gi
----
+
The following example changes the PVC size to 20 gigabytes for Thanos Ruler:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    thanosRuler:
      volumeClaimTemplate:
        spec:

>           storageClassName: *local-storage*

Reviewer 2: storageClassName: local-storage

          resources:
            requests:
              storage: 20Gi
----
+
[NOTE]
====
Storage requirements for the `thanosRuler` component depend on the number of rules that are evaluated and how many samples each rule generates.
====

. Save the file to apply the changes. The pods affected by the new configuration restart automatically.
+
[WARNING]
====
When you save changes to a monitoring config map, the pods and other resources in the related project might be redeployed. The running monitoring processes in that project might also be restarted.
====

. Manually patch every PVC with the updated storage request. The following example resizes the storage size for the `openshift-monitoring` Prometheus component to 100Gi:
+
[source,terminal]
----
# for p in $(oc get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \

>   oc patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}'; \

Reviewer 2: PVC is restricted to namespace,
```
# for p in $(oc -n openshift-monitoring get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \
  oc -n openshift-monitoring patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}'; 
```

Author: Updated per your suggestion

  done
----

> . Delete the underlying StatefulSet with the `--cascade=orphan` parameter:

Reviewer 2: we could delete this step, reason
1. there is not `--cascade` flag now
```
# oc version
Client Version: 4.12.0-0.nightly-2022-10-05-053337
Kustomize Version: v4.5.4
Server Version: 4.12.0-0.nightly-2022-11-10-033725
Kubernetes Version: v1.25.2+f33d98e

# oc -n openshift-monitoring get statefulset -l app.kubernetes.io/name=prometheus --cascade=orphan
error: unknown flag: --cascade
See 'oc get --help' for usage.
```
2. old pods would be removed automatically and new pods would be restarted automatically

Reviewer: I think you typed the wrong command by accident, you used `get` instead of `delete`, I'm not aware of `--cascade` being deprecated

Author: @juzhao @JoaoBraveCoding - Can you please provide guidance on whether or not I need to change anything? Does `--cascade` need to be removed from the command? Thanks!

Reviewer 2: we could remove the step
`Delete the underlying StatefulSet with the --cascade=orphan parameter:`
reason please see my previous comment

Reviewer 3: I think we do really need this step. This step is specifically called out in the [prometheus-operator docs](https://prometheus-operator.dev/docs/operator/storage/#resizing-volumes) and iiuc is caused by a missing k8s feature.

And @JoaoBraveCoding  is correct, the cascade option should still be there. This is 4.11 but 4.12 should look the same:
```
jan@jflaptop ~/Downloads % oc version
Client Version: 4.11.13
Kustomize Version: v4.5.4
jan@jflaptop ~/Downloads % oc delete --help | grep cascade -C2
	If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.

    --cascade='background':
	Must be "background", "orphan", or "foreground". Selects the deletion cascading strategy for the dependents (e.g. Pods created by a ReplicationController). Defaults to background.
```

Reviewer 2: ok, my mistake, I used `oc get` command, which should be `oc delete`, no need to remove, we need this step, lgtm
```
# oc version
Client Version: 4.12.0-0.nightly-2022-10-05-053337
Kustomize Version: v4.5.4
Server Version: 4.12.0-0.nightly-2022-11-10-033725

# oc -n openshift-monitoring delete statefulset -l app.kubernetes.io/name=prometheus --cascade=orphan
statefulset.apps "prometheus-k8s" deleted
```

Author: Thanks, all, for your input and thorough review comments. Very much appreciated.

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    thanosRuler:
      volumeClaimTemplate:
        spec:
          storageClassName: local-storage
          resources:
            requests:
              storage: 20Gi
----
+
[NOTE]
====
Storage requirements for the `thanosRuler` component depend on the number of rules that are evaluated and how many samples each rule generates.
====

. Save the file to apply the changes. The pods affected by the new configuration restart automatically.
+
[WARNING]
====
When you save changes to a monitoring config map, the pods and other resources in the related project might be redeployed. The running monitoring processes in that project might also be restarted.
====

. Manually patch every PVC with the updated storage request. The following example resizes the storage size for the Prometheus component in the `openshift-monitoring` namespace to 100Gi:
+
[source,terminal]
----
# for p in $(oc -n openshift-monitoring get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \

>   oc -n openshift-monitoring patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}'; 

Reviewer 2: missed `done` at the end
```
# for p in $(oc -n openshift-monitoring get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \
  oc -n openshift-monitoring patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}';  done
```

Author: fixed

// Module included in the following assemblies:
//
// * monitoring/configuring-the-monitoring-stack.adoc

:_content-type: PROCEDURE
[id="resizing-a-persistent-storage-volume_{context}"]
= Resizing a persistent storage volume

> {product-title} does not support resizing an existing persistent storage volume using StatefulSets, even if the underlying `StorageClass` used supports persistent volume sizing.

Author: I updated both to add `resources` per your suggestion.

> {product-title} does not support resizing an existing persistent storage volume using `StatefulSet` resources, even if the underlying `StorageClass` resource used supports persistent volume sizing.

Author: changed per your suggestion

Therefore, even if you update the `storage` field for an existing persistent volume claim (PVC) with a larger size, this setting will not be propagated to the associated persistent volume (PV).

> However, resizing a PV is still possible by using a manual process. If you want to resize a PV for a monitoring component such as Prometheus, Thanos Ruler, or Alertmanager, you can update the appropriate config map where the component is configured, patch the PVC, and then delete and orphan the pods.

Author: I think two sentences is easier to parse. So I've followed your suggestion.

> Orphaning the pods recreates the StatefulSet immediately and automatically updates the size of the volumes mounted in the pods with the new PVC settings.

Author: Added in `resource`

No service disruption occurs during this process. 

.Prerequisites

* You have installed the OpenShift CLI (`oc`).
* *If you are configuring core {product-title} monitoring components*:
** You have access to the cluster as a user with the `cluster-admin` role.
** You have created the `cluster-monitoring-config` `ConfigMap` object.
** You have configured at least one PVC for core {product-title} monitoring components.
* *If you are configuring components that monitor user-defined projects*:
** You have access to the cluster as a user with the `cluster-admin` role, or as a user with the `user-workload-monitoring-config-edit` role in the `openshift-user-workload-monitoring` project.
** You have created the `user-workload-monitoring-config` `ConfigMap` object.
** You have configured at least one PVC for components that monitor user-defined projects.

.Procedure

. Edit the `ConfigMap` object:
** *To resize a PVC for a component that monitors core {product-title} projects*:
.. Edit the `cluster-monitoring-config` `ConfigMap` object in the `openshift-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-monitoring edit configmap cluster-monitoring-config
----

> .. Add a new storage size for the PVC configuration for the component under `data/config.yaml`:

Author: Leaving as-is

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    <component>: <1>
      volumeClaimTemplate:
        spec:
          storageClassName: <storage_class> <2>
          resources:
            requests:
              storage: <amount_of_storage> <3>
----
<1> Specify the core monitoring component.
<2> Specify the storage class.
<3> Specify the new size for the storage volume.
+

> The following example configures a PVC that changes the local persistent storage to 100 gigabytes for the Prometheus instance that monitors core {product-title} components:

Author: changed to `configures a PVC that sets the local persistent storage to 100 gigabytes` to avoid repeating the word `configures`

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      volumeClaimTemplate:
        spec:
          storageClassName: local-storage
          resources:
            requests:
              storage: 100Gi
----
+
The following example configures a PVC that changes local persistent storage for Alertmanager to 40 gigabytes:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    alertmanagerMain:
      volumeClaimTemplate:
        spec:

>           storageClassName: local-storage

Author: The preview hasn't updated, I think. An earlier version incorrectly had asterisks, and I removed them.

          resources:
            requests:
              storage: 40Gi
----

** *To resize a PVC for a component that monitors user-defined projects*:
+
[NOTE]
====
You can resize the volumes for the Thanos Ruler and Prometheus instances that monitor user-defined projects.
====
+
.. Edit the `user-workload-monitoring-config` `ConfigMap` object in the `openshift-user-workload-monitoring` project:
+
[source,terminal]
----
$ oc -n openshift-user-workload-monitoring edit configmap user-workload-monitoring-config
----

.. Update the PVC configuration for the monitoring component under `data/config.yaml`:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    <component>: <1>
      volumeClaimTemplate:
        spec:
          storageClassName: <storage_class> <2>
          resources:
            requests:
              storage: <amount_of_storage> <3>
----
<1> Specify the core monitoring component.
<2> Specify the storage class.
<3> Specify the new size for the storage volume.
+

> The following example changes the PVC size to 100 gigabytes for the Prometheus instance that monitors user-defined projects:

Author: changed to `configures` per your suggestion

+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    prometheus:
      volumeClaimTemplate:
        spec:
          storageClassName: local-storage
          resources:
            requests:
              storage: 100Gi
----
+
The following example changes the PVC size to 20 gigabytes for Thanos Ruler:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: user-workload-monitoring-config
  namespace: openshift-user-workload-monitoring
data:
  config.yaml: |
    thanosRuler:
      volumeClaimTemplate:
        spec:
          storageClassName: local-storage
          resources:
            requests:
              storage: 20Gi
----
+
[NOTE]
====
Storage requirements for the `thanosRuler` component depend on the number of rules that are evaluated and how many samples each rule generates.
====

. Save the file to apply the changes. The pods affected by the new configuration restart automatically.
+
[WARNING]
====

> When you save changes to a monitoring config map, the pods and other resources in the related project might be redeployed. The running monitoring processes in that project might also be restarted.

Author: Good suggestion. Made the change. Thanks!

====

. Manually patch every PVC with the updated storage request. The following example resizes the storage size for the Prometheus component in the `openshift-monitoring` namespace to 100Gi:
+
[source,terminal]
----
# for p in $(oc -n openshift-monitoring get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \
  oc -n openshift-monitoring patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}'; \
  done

----

. Delete the underlying StatefulSet with the `--cascade=orphan` parameter:
+
[source,terminal]
----

> # oc delete statefulset -l app.kubernetes.io/name=prometheus --cascade=orphan

Author: good catch. `oc` commands don't require sudo permissions. Fixed.
